aten/src/ATen/cpu/vec256/vsx/vec256_common_vsx.h

```diff
@@ -0,0 +1,216 @@
+#pragma once
+
+#include <ATen/cpu/vec256/intrinsics.h>
+#include <ATen/cpu/vec256/vec256_base.h>
+#include <ATen/cpu/vec256/vsx/vsx_helpers.h>
+#include <ATen/cpu/vec256/vsx/vec256_double_vsx.h>
+#include <ATen/cpu/vec256/vsx/vec256_float_vsx.h>
+#include <ATen/cpu/vec256/vsx/vec256_int16_vsx.h>
+#include <ATen/cpu/vec256/vsx/vec256_int32_vsx.h>
+#include <ATen/cpu/vec256/vsx/vec256_int64_vsx.h>
+#include <ATen/cpu/vec256/vsx/vec256_qint32_vsx.h>
+#include <ATen/cpu/vec256/vsx/vec256_qint8_vsx.h>
+#include <ATen/cpu/vec256/vsx/vec256_quint8_vsx.h>
+#include <ATen/cpu/vec256/vsx/vec256_complex_float_vsx.h>
+#include <ATen/cpu/vec256/vsx/vec256_complex_double_vsx.h>
+namespace at {
+namespace vec256 {
+
+namespace {
+
+DEFINE_CLAMP_FUNCS(c10::quint8)
+DEFINE_CLAMP_FUNCS(c10::qint8)
+DEFINE_CLAMP_FUNCS(c10::qint32)
+DEFINE_CLAMP_FUNCS(int16_t)
+DEFINE_CLAMP_FUNCS(int32_t)
+DEFINE_CLAMP_FUNCS(int64_t)
+DEFINE_CLAMP_FUNCS(float)
+DEFINE_CLAMP_FUNCS(double)
+
+template <>
+Vec256<double> inline __inline_attrs fmadd(
+    const Vec256<double>& a,
+    const Vec256<double>& b,
+    const Vec256<double>& c) {
+  return Vec256<double>{
+      vec_madd(a.vec0(), b.vec0(), c.vec0()),
+      vec_madd(a.vec1(), b.vec1(), c.vec1())};
+}
+
+template <>
+Vec256<int64_t> inline __inline_attrs fmadd(
+    const Vec256<int64_t>& a,
+    const Vec256<int64_t>& b,
+    const Vec256<int64_t>& c) {
+  return Vec256<int64_t>{
+      a.vec0() * b.vec0() + c.vec0(), a.vec1() * b.vec1() + c.vec1()};
+}
+template <>
+Vec256<int32_t> inline __inline_attrs fmadd(
+    const Vec256<int32_t>& a,
+    const Vec256<int32_t>& b,
+    const Vec256<int32_t>& c) {
+  return Vec256<int32_t>{
+      a.vec0() * b.vec0() + c.vec0(), a.vec1() * b.vec1() + c.vec1()};
+}
+template <>
+Vec256<int16_t> inline __inline_attrs fmadd(
+    const Vec256<int16_t>& a,
+    const Vec256<int16_t>& b,
+    const Vec256<int16_t>& c) {
+  return Vec256<int16_t>{
+      a.vec0() * b.vec0() + c.vec0(), a.vec1() * b.vec1() + c.vec1()};
+}
+
+DEFINE_REINTERPRET_CAST_TO_ALL_FUNCS(float)
+DEFINE_REINTERPRET_CAST_TO_ALL_FUNCS(double)
+DEFINE_REINTERPRET_CAST_TO_ALL_FUNCS(int64_t)
+DEFINE_REINTERPRET_CAST_TO_ALL_FUNCS(int32_t)
+DEFINE_REINTERPRET_CAST_TO_ALL_FUNCS(int16_t)
+
+template <>
+Vec256<int64_t> inline __inline_attrs convert_to_int_of_same_size<double>(
+    const Vec256<double>& src) {
+  return Vec256<int64_t>{vec_signed(src.vec0()), vec_signed(src.vec1())};
+}
+
+template <>
+Vec256<int32_t> inline __inline_attrs convert_to_int_of_same_size<float>(
+    const Vec256<float>& src) {
+  return Vec256<int32_t>{vec_signed(src.vec0()), vec_signed(src.vec1())};
+}
+
+template <>
+inline void convert(const int32_t* src, float* dst, int64_t n) {
+  // int32_t and float have same size
+  size_t N = n;
+  size_t NE = N & (~(Vec256<int32_t>::size() - 1));
```

Comment:
Do you mean

    NE = N - Vec256<int32_t>::size();

? Is there a reason to use `&` here instead of the more straightforward `-`?

Reply:
I used it for the tail.  but thats right I could go with more obvious one by defining size_t i outside